lib/mqtt/client.rb

```diff
@@ -329,7 +331,19 @@ def publish(topic, payload='', retain=false, qos=0)
     )
 
     # Send the packet
-    send_packet(packet)
+    res = send_packet(packet)
+
+    if packet.qos > 0
+      Timeout.timeout(@ack_timeout) do
+        while connected? do
+          @pubacks_semaphore.synchronize do
+            return res if @pubacks.delete(packet.id)
+          end
+          sleep 0.1
```

Comment:
Sorry for the delay in reviewing this PR.

I am not very sure about the sleep 0.1. A tenth of a second is quite a long time to sleep.

Ideally this would have some kind of event mechanism to perform notification between threads (I have looked at using pipes and select).

But maybe for now just make the sleep shorter?